src/Microsoft.AspNetCore.Mvc.Core/Internal/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    /// <summary>
+    /// This class represents a general purpose executor for a method(static/instance) on an object using MethodInfo.
+    /// This executor is an alternative approach for MethodInfo.Invoke(). 
+    /// Given the MethodInfo and the type of the target object, it creates a delegate and calls the method using it.
+    /// The delegate is created dynamically using the Linq expressions. 
+    /// The executor should be cached to get the maximum performance benefit or else compiling Linq expression adds overhead.
+    /// </summary>
+    public class ObjectMethodExecutor
+    {
+        private ActionExecutor _executor;
+
+        public ObjectMethodExecutor(MethodInfo methodInfo, TypeInfo targetTypeInfo)
+        {            
+            if (methodInfo == null)
+            {
+                throw new ArgumentNullException(nameof(methodInfo));
+            }
+
+            if (targetTypeInfo == null)
+            {
+                throw new ArgumentNullException(nameof(targetTypeInfo));
+            }
+
+            _executor = GetExecutor(methodInfo, targetTypeInfo);
+            MethodInfo = methodInfo;
+        }
+
+        private delegate object ActionExecutor(object target, object[] parameters);
+
+        private delegate void VoidActionExecutor(object target, object[] parameters);
+
+        public MethodInfo MethodInfo { get; private set; }
+
+        /// <inheritdoc />
+        public object Execute(object target, object[] parameters)
+        {
+            return _executor(target, parameters);
+        }
+
+        private static ActionExecutor GetExecutor(MethodInfo methodInfo, TypeInfo targetTypeInfo)
+        {
+            // Parameters to executor
+            ParameterExpression targetParameter = Expression.Parameter(typeof(object), "target");
+            ParameterExpression parametersParameter = Expression.Parameter(typeof(object[]), "parameters");
```

Comment:
I know this is a design PR, but make sure to do a pass here for style. (use var, etc).